tests/Constraint/ConstraintTest.php

```diff
@@ -91,29 +169,159 @@ public function testVersionMatchSucceeds($requireOperator, $requireVersion, $pro
-            array('==', 'dev-foo-bist', '==', 'dev-foo-aist'),
-            array('<=', 'dev-foo-bist', '>=', 'dev-foo-aist'),
-            array('>=', 'dev-foo-bist', '<', 'dev-foo-aist'),
-            array('<', '0.12', '==', 'dev-foo'), // branches are not comparable
-            array('>', '0.12', '==', 'dev-foo'), // branches are not comparable
+
+            // different branch names
+            array('==', 'dev-foo-bar', '==', 'dev-foo-xyz'),
+            array('==', 'dev-foo-bar', '<', 'dev-foo-xyz'),
+            array('==', 'dev-foo-bar', '<=', 'dev-foo-xyz'),
+            array('==', 'dev-foo-bar', '>=', 'dev-foo-xyz'),
+            array('==', 'dev-foo-bar', '>', 'dev-foo-xyz'),
+
+            array('<', 'dev-foo-bar', '==', 'dev-foo-xyz'),
+            array('<', 'dev-foo-bar', '<', 'dev-foo-xyz'),
+            array('<', 'dev-foo-bar', '<=', 'dev-foo-xyz'),
+            array('<', 'dev-foo-bar', '>=', 'dev-foo-xyz'),
+            array('<', 'dev-foo-bar', '>', 'dev-foo-xyz'),
+            array('<', 'dev-foo-bar', '!=', 'dev-foo-xyz'),
+
+            array('<=', 'dev-foo-bar', '==', 'dev-foo-xyz'),
+            array('<=', 'dev-foo-bar', '<', 'dev-foo-xyz'),
+            array('<=', 'dev-foo-bar', '<=', 'dev-foo-xyz'),
+            array('<=', 'dev-foo-bar', '>=', 'dev-foo-xyz'),
+            array('<=', 'dev-foo-bar', '>', 'dev-foo-xyz'),
+            array('<=', 'dev-foo-bar', '!=', 'dev-foo-xyz'),
+
+            array('>=', 'dev-foo-bar', '==', 'dev-foo-xyz'),
+            array('>=', 'dev-foo-bar', '<', 'dev-foo-xyz'),
+            array('>=', 'dev-foo-bar', '<=', 'dev-foo-xyz'),
+            array('>=', 'dev-foo-bar', '>=', 'dev-foo-xyz'),
+            array('>=', 'dev-foo-bar', '>', 'dev-foo-xyz'),
+            array('>=', 'dev-foo-bar', '!=', 'dev-foo-xyz'),
+
+            array('>', 'dev-foo-bar', '==', 'dev-foo-xyz'),
+            array('>', 'dev-foo-bar', '<', 'dev-foo-xyz'),
+            array('>', 'dev-foo-bar', '<=', 'dev-foo-xyz'),
+            array('>', 'dev-foo-bar', '>=', 'dev-foo-xyz'),
+            array('>', 'dev-foo-bar', '>', 'dev-foo-xyz'),
+            array('>', 'dev-foo-bar', '!=', 'dev-foo-xyz'),
+
+            // same branch names
+            array('==', 'dev-foo-bar', '<', 'dev-foo-bar'),
+            array('==', 'dev-foo-bar', '<=', 'dev-foo-bar'),
+            array('==', 'dev-foo-bar', '>=', 'dev-foo-bar'),
+            array('==', 'dev-foo-bar', '>', 'dev-foo-bar'),
+            array('==', 'dev-foo-bar', '!=', 'dev-foo-bar'),
+
+            array('<', 'dev-foo-bar', '==', 'dev-foo-bar'),
+            array('<', 'dev-foo-bar', '<', 'dev-foo-bar'),
+            array('<', 'dev-foo-bar', '<=', 'dev-foo-bar'),
+            array('<', 'dev-foo-bar', '>=', 'dev-foo-bar'),
+            array('<', 'dev-foo-bar', '>', 'dev-foo-bar'),
+            array('<', 'dev-foo-bar', '!=', 'dev-foo-bar'),
+
+            array('<=', 'dev-foo-bar', '==', 'dev-foo-bar'),
```

Comment:
shouldn't it be OK?

Reply:
Well the idea here is to generally return false on any operator other than == / != for branches. It's just confusing if <= dev-foo-bar works, but < dev-foo-bar has no meaning. I was going back and forth on this too, but I think it's clearer to just make everything relating to branches return false if improper operators are used.